GetStream/stream-chat-swift
# [CIS-2241] Do not increment channel unread count for already processed messages

### 🔗 Issue Links

https://stream-io.atlassian.net/browse/CIS-2241

### 🎯 Goal

Fix inconsistencies for channel unread counts

### 📝 Summary

There are edge cases in which we get channel updates through HTTP requests, but we then get events for new messages that we have already processed. In this cases, we might end up double counting those messages, and showing an incorrect unread count for those

### 🛠 Implementation

With this changes, for each WebSocket event that notifies of new messages, we are going to be checking if the message is already in database. If the message is already in our database, it means it was already processed, and therefore we don't need to update the unread count based on the event.
For the messages that we receive through WebSockets that are not yet in the database, the behaviour is kept the same. 

### 🧪 Manual Testing Notes

The best way to reproduce this in real world is:
1. Open the app in device A, wait until the sync is complete
2. Stay in the channel list
3. From another device (B), send messages to A, and observe how the unread counts work properly
4. Send A to background (make sure `staysConnectedInBackground` is set to `false`)
5. Open app in device A

Expected behaviour:
6. We fetch all the content, and the unread counts are kept the same

Old behaviour:
7. The messages that were sent during the previous session are DOUBLE COUNTED.
8. In most of the cases the unread count goes back to normal, but in some exceptions the wrong count was kept.

### ☑️ Contributor Checklist


### 🎁 Meme

![](https://media.giphy.com/media/tZrbdhZCsdORMItXrz/giphy.gif)

[CIS-2241]: https://stream-io.atlassian.net/browse/CIS-2241?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### Notes

Ended up going for the count (vs fetch) approach.
// 200 objects ----
// Count:   0,00167 0,00199 0,0006
// Fetch:   0,00129 0,00189 0,0017

// 1000 objects ---
// Count:   0,0006 0,0007
// Fetch:   0,0007 0,0006
 Faster when having ~200, and same when as high as 1000.
In any case, in most of the cases when using the /sync endpoint we would not even reach 200
